### ISSUE TITLE
Support multiple output formats

### ISSUE THREAD

**Issue by Nickersoft** [TheOpenDictionary/odict#59]
Something I've been thinking about lately is the ability to serialize ODict dictionary dumps and entries to different formats. Right now, dumps are only in XML and entries are only in JSON. I feel like these should be configurable via a flag, such as `-f`. For dumps, XML can be the default, and for entries, we can pretty-print the output to stdout.

@maxdemaio I also think SQL could be a good output format, and would make seeding [web](https://github.com/TheOpenDictionary/web) much, _much_ easier. The table names would be opinionated, but we could just make sure the schemas match between this repo and web. Happy to pair on this if you feel like learning Go lol. If not, I can pick up this issue and we can leverage it to add more dictionaries to the db.

**Comment by maxdemaio** [comment]
Would totally be down to pair on this one. I really like this idea. Also, it would bridge the gap between the two if web used the SQL structure outputted from this repo. We could hop on a Discord call or just hash it out over Twitter and take stabs at it sometime 💪🏻

**Comment by maxdemaio** [comment]
Created the branch `59-support-multiple-output-formats` for future work 👍🏻

**Comment by Nickersoft** [comment]
Ah yeah this is perfect! We can use a `-f` to specify output format. I recently added `-f` support [for lookups](https://github.com/TheOpenDictionary/odict/blob/59-support-multiple-output-formats/cli/lookup.go#L15), and we can add a [converter](https://github.com/TheOpenDictionary/odict/blob/59-support-multiple-output-formats/go/converters.go) that dumps it to SQL.

**Comment by maxdemaio** [comment]
> Ah yeah this is perfect! We can use a `-f` to specify output format. I recently added `-f` support [for lookups](https://github.com/TheOpenDictionary/odict/blob/59-support-multiple-output-formats/cli/lookup.go#L15), and we can add a [converter](https://github.com/TheOpenDictionary/odict/blob/59-support-multiple-output-formats/go/converters.go) that dumps it to SQL.

Sweet! That makes sense to me. I feel like specifying the flag shouldn't be hard until we get to the SQL part 😅  Something like this? Then I guess we could have a struct of valid formats:

```
odict d -f sql mydictionary.odict output.sql
```

```
odict d -f xml mydictionary.odict output.xml
```

and default would be:

```
odict d mydictionary.odict output.xml
```

**Comment by Nickersoft** [comment]
Yep yep! I like it. Yeah, unfortunately, unlike JSON & XML, we can't use any built-in Go magic marshaling haha. Taking a quick look, however, it looks like the [bob](https://github.com/stephenafamo/bob) package could be a fast & lightweight way to use a query builder for SQL. It'd also open the door to allowing us to support different SQL flavors down the line. You'd most likely just have to loop through all the etymologies/usages/etc. manually and add them to the query builder.

**Comment by maxdemaio** [comment]
> Yep yep! I like it. Yeah, unfortunately, unlike JSON & XML, we can't use any built-in Go magic marshaling haha. Taking a quick look, however, it looks like the [bob](https://github.com/stephenafamo/bob) package could be a fast & lightweight way to use a query builder for SQL. It'd also open the door to allowing us to support different SQL flavors down the line. You'd most likely just have to loop through all the etymologies/usages/etc. manually and add them to the query builder.

I like this a lot to avoid being hard stuck on one SQL flavor. This will also let other devs using ODict to develop freely. This feature has some huge impact 🔥

Once this is ready it's pretty much "all systems go" for our `web` repo. I don't think we'd even need a seed file and we'd just use this method to initialize a DB and let Prisma handle the rest 👀

**Comment by Nickersoft** [comment]
Do you think you have any time to take a first stab at it? My workload is stretched pretty thin at the moment, but would love to see this – otherwise I'll just carve out time on some weekend to get a POC together.

**Comment by maxdemaio** [comment]
Started working on this on https://github.com/TheOpenDictionary/odict/tree/59-support-multiple-output-formats!

So far, all the routing from the CLI is done so you can do:

- `odict d -f sql mydictionary.odict output.sql` (brings you to a function for SQL dumps)
- `odict d -f xml mydictionary.odict output.xml` (brings you to the xml function for SQL dumps)
- `odict d -f mydictionary.odict output.xml` (brings you to the xml function for SQL dumps)

As I explore Bob and other packages, I realize we'd have to come up with a schema for the DB. Luckily we've already done this in `web` here: https://github.com/TheOpenDictionary/web/blob/main/prisma/schema.prisma. As you mentioned before this would be the opinionated part.

After the schema is made, we just would have to spit out a bunch of insert statements (and making sure there's no sql injections randomly in the dict data). This is where something like Bob could come in to make sure these insert statements work for different SQL flavors.

I'm a little stuck on how we'd define the schema since I don' think Bob does this and states, "Bob is a 'database-first' ORM. That means you must first create your database schema." In web, we use Prisma to do it for us but that's NodeJS based so can't use it on the Go side of things.

**Comment by Nickersoft** [comment]
Wow this looks great! Yep, that approach sounds right to me. You have a good point – maybe a workaround is making the output format more specific? Like `pgsql` instead of `sql`?

I also found some other interesting projects
- [meddler](https://github.com/russross/meddler) is a seemingly unmaintained project that converts between SQL and Go structs using serializers.
- [GoORM](https://gorm.io) is supposed to be used with databases, _but_ it looks like it _does_ [support raw query building](https://gorm.io/docs/sql_builder.html) _and_ has built-in support for [serializing structs](https://gorm.io/docs/serializer.html). Could be helpful.

**Comment by maxdemaio** [comment]
I think GoORM would be a good choice. We could attach on some SQL tags just like you did for JSON like:

```go
type Dictionary struct {
	ID                  int64 `gorm:"column:id;primary_key;AUTO_INCREMENT"`
	Name                string `gorm:"column:name"`
	SourceLanguageID    int64 `gorm:"column:source_language_id"`
	TargetLanguageID    int64 `gorm:"column:target_language_id"`
	SourceLanguage      Language `gorm:"ForeignKey:SourceLanguageID"`
	TargetLanguage      Language `gorm:"ForeignKey:TargetLanguageID"`
	Entries             []Entry
}

type Entry struct {
	ID                  int64 `gorm:"column:id;primary_key;AUTO_INCREMENT"`
	CreatedAt           time.Time `gorm:"column:created_at;DEFAULT:CURRENT_TIMESTAMP"`
	UpdatedAt           time.Time `gorm:"column:updated_at;DEFAULT:CURRENT_TIMESTAMP"`
	Term                string `gorm:"column:term"`
	Etymologies         []Etymology `gorm:"ForeignKey:EntryID"`
	DictionaryID        int64 `gorm:"column:dictionary_id"`
	Dictionary          Dictionary `gorm:"ForeignKey:DictionaryID"`
}
```

🤔 I'll see how far I can get with something like this

**Comment by maxdemaio** [comment]
Per ChatGPT,

"We can use the Go structs to generate the create statements without a DB like so:

- Write the Go structs that represent your database tables and their relationships using the GORM library. In your example, you have already defined the DictionarySqlTable and EntrySqlTable structs.
- Write a code generator that generates the SQL statements for creating the database tables based on the Go structs. You can use the GORM library to retrieve metadata about the Go structs and use that to generate the SQL statements. This code generator can be written in any programming language of your choice.
- Add a //go:generate directive to your Go source file that specifies the command to run your code generator. This directive should be placed at the top of the file, before the package declaration.
- Run the go generate command to execute the code generator and generate the SQL statements for creating the database tables.
- Use the generated SQL statements to create the database tables and insert data into them using your database management tool of choice.

The go generate command runs the code generator specified in the `//go:generate` directive, which generates the SQL statements for creating the database tables based on the Go structs. We can then use the generated SQL statements to create the database tables and insert data into them.

Note that the go:generate directive is a build-time tool that generates code, so it is not executed when you run your Go program. Instead, you need to explicitly run the go generate command to generate the code."

I am going to investigate along these lines and see what I come up with 👍🏻

**Comment by Nickersoft** [comment]
Yeahhh I would take what it says with a grain of salt haha. Not sure why it's recommending a go directive – I've never even used them lol.

**Comment by maxdemaio** [comment]
Lol yeah I had a feeling that wasn't the right way to go (no pun intended). 

Watched a quick presentation on a library called "sq" developed by a cool guy named @bokwoon95 https://youtu.be/C2JfH9FDusQ. Looks like there is a to [use table structs as a declaritive schema](https://bokwoon.neocities.org/sq#toc-declarative-schema). Will look into this next 👍🏻

**Comment by bokwoon95** [comment]
The code is an example of what you're trying to achieve (put it in `main.go` and run it with `go run main.go`). 

I'm using Go's [embed feature](https://gobyexample.com/embed-directive) to bake the source code straight into the Go binary so it doesn't depend on any external file.

You should also pull in the latest version of sqddl (`go install -tags=fts5 github.com/bokwoon95/sqddl@latest`), I made a bunch of changes to make it work.

```go
package main

import (
	"embed"
	"fmt"

	"github.com/bokwoon95/sq"
	"github.com/bokwoon95/sqddl/ddl"
)

//go:embed main.go
var fsys embed.FS

func main() {
	dialects := []string{sq.DialectSQLite, sq.DialectPostgres, sq.DialectMySQL, sq.DialectSQLServer}
	for _, dialect := range dialects {
		fmt.Println("\n--[ " + dialect + " ]--\n")
		generateCmd := &ddl.GenerateCmd{
			Dialect:   dialect,
			DirFS:     fsys,
			Filenames: []string{"main.go"},
			DryRun:    true,
		}
		err := generateCmd.Run()
		if err != nil {
			fmt.Println(err)
		}
	}
}

type LANGUAGE struct {
	sq.TableStruct
	ID   sq.NumberField `ddl:"type=BIGINT primarykey auto_increment autoincrement identity"`
	CODE sq.StringField `ddl:"notnull unique"`
	FLAG sq.StringField `ddl:"notnull unique"`
}

type DICTIONARY struct {
	sq.TableStruct
	ID                 sq.NumberField `ddl:"type=BIGINT primarykey auto_increment autoincrement identity"`
	NAME               sq.StringField `ddl:"notnull"`
	SOURCE_LANGUAGE_ID sq.NumberField `ddl:"type=BIGINT notnull references=language.id"`
	TARGET_LANGUAGE_ID sq.NumberField `ddl:"type=BIGINT notnull references=language.id"`
}

type ENTRY struct {
	sq.TableStruct
	ID            sq.NumberField `ddl:"type=BIGINT primarykey auto_increment autoincrement identity"`
	CREATED_AT    sq.TimeField   `ddl:"default=CURRENT_TIMESTAMP"`
	UPDATED_AT    sq.TimeField
	TERM          sq.StringField `ddl:"notnull"`
	DICTIONARY_ID sq.NumberField `ddl:"type=BIGINT references=dictionary.id"`
}

type ETYMOLOGY struct {
	sq.TableStruct
	ID          sq.StringField `ddl:"primarykey"` // cuid() generation is handled by prisma, not the database.
	DESCRIPTION sq.StringField
	ENTRY_ID    sq.NumberField `ddl:"type=BIGINT notnull references=entry.id"`
}

type USAGE struct {
	sq.TableStruct
	ID           sq.NumberField `ddl:"type=BIGINT primarykey auto_increment autoincrement identity"`
	ETYMOLOGY_ID sq.StringField `ddl:"notnull references=etymology.id"`
}

type GROUP struct {
	sq.TableStruct
	ID          sq.NumberField `ddl:"type=BIGINT primarykey auto_increment autoincrement identity"`
	DESCRIPTION sq.StringField
	USAGE_ID    sq.NumberField `ddl:"type=BIGINT notnull references=usage.id"`
}

type DEFINITION struct {
	sq.TableStruct
	ID       sq.NumberField `ddl:"type=BIGINT primarykey auto_increment autoincrement identity"`
	TEXT     sq.StringField `ddl:"notnull"`
	USAGE_ID sq.NumberField `ddl:"type=BIGINT references=usage.id"`
	GROUP_ID sq.NumberField `ddl:"type=BIGINT references=group.id"`
}

type EXAMPLE struct {
	sq.TableStruct
	ID            sq.NumberField `ddl:"type=BIGINT primarykey auto_increment autoincrement identity"`
	TEXT          sq.StringField `ddl:"notnull"`
	DEFINITION_ID sq.NumberField `ddl:"type=BIGINT notnull references=definition.id"`
}
```
Output:
```sql

--[ sqlite ]--

-- 20230221194111.sql
CREATE TABLE language (
    id INTEGER PRIMARY KEY AUTOINCREMENT
    ,code TEXT NOT NULL
    ,flag TEXT NOT NULL

    ,CONSTRAINT language_code_key UNIQUE (code)
    ,CONSTRAINT language_flag_key UNIQUE (flag)
);

CREATE TABLE dictionary (
    id INTEGER PRIMARY KEY AUTOINCREMENT
    ,name TEXT NOT NULL
    ,source_language_id BIGINT NOT NULL
    ,target_language_id BIGINT NOT NULL

    ,CONSTRAINT dictionary_source_language_id_fkey FOREIGN KEY (source_language_id) REFERENCES language (id)
    ,CONSTRAINT dictionary_target_language_id_fkey FOREIGN KEY (target_language_id) REFERENCES language (id)
);

CREATE TABLE entry (
    id INTEGER PRIMARY KEY AUTOINCREMENT
    ,created_at DATETIME DEFAULT CURRENT_TIMESTAMP
    ,updated_at DATETIME
    ,term TEXT NOT NULL
    ,dictionary_id BIGINT

    ,CONSTRAINT entry_dictionary_id_fkey FOREIGN KEY (dictionary_id) REFERENCES dictionary (id)
);

CREATE TABLE etymology (
    id TEXT PRIMARY KEY NOT NULL
    ,description TEXT
    ,entry_id BIGINT NOT NULL

    ,CONSTRAINT etymology_entry_id_fkey FOREIGN KEY (entry_id) REFERENCES entry (id)
);

CREATE TABLE usage (
    id INTEGER PRIMARY KEY AUTOINCREMENT
    ,etymology_id TEXT NOT NULL

    ,CONSTRAINT usage_etymology_id_fkey FOREIGN KEY (etymology_id) REFERENCES etymology (id)
);

CREATE TABLE "group" (
    id INTEGER PRIMARY KEY AUTOINCREMENT
    ,description TEXT
    ,usage_id BIGINT NOT NULL

    ,CONSTRAINT group_usage_id_fkey FOREIGN KEY (usage_id) REFERENCES usage (id)
);

CREATE TABLE definition (
    id INTEGER PRIMARY KEY AUTOINCREMENT
    ,text TEXT NOT NULL
    ,usage_id BIGINT
    ,group_id BIGINT

    ,CONSTRAINT definition_usage_id_fkey FOREIGN KEY (usage_id) REFERENCES usage (id)
    ,CONSTRAINT definition_group_id_fkey FOREIGN KEY (group_id) REFERENCES "group" (id)
);

CREATE TABLE example (
    id INTEGER PRIMARY KEY AUTOINCREMENT
    ,text TEXT NOT NULL
    ,definition_id BIGINT NOT NULL

    ,CONSTRAINT example_definition_id_fkey FOREIGN KEY (definition_id) REFERENCES definition (id)
);

--[ postgres ]--

-- 20230221194111_01_tables.sql
CREATE TABLE language (
    id BIGINT NOT NULL GENERATED BY DEFAULT AS IDENTITY
    ,code TEXT NOT NULL
    ,flag TEXT NOT NULL

    ,CONSTRAINT language_id_pkey PRIMARY KEY (id)
    ,CONSTRAINT language_code_key UNIQUE (code)
    ,CONSTRAINT language_flag_key UNIQUE (flag)
);

CREATE TABLE dictionary (
    id BIGINT NOT NULL GENERATED BY DEFAULT AS IDENTITY
    ,name TEXT NOT NULL
    ,source_language_id BIGINT NOT NULL
    ,target_language_id BIGINT NOT NULL

    ,CONSTRAINT dictionary_id_pkey PRIMARY KEY (id)
);

CREATE TABLE entry (
    id BIGINT NOT NULL GENERATED BY DEFAULT AS IDENTITY
    ,created_at TIMESTAMPTZ DEFAULT CURRENT_TIMESTAMP
    ,updated_at TIMESTAMPTZ
    ,term TEXT NOT NULL
    ,dictionary_id BIGINT

    ,CONSTRAINT entry_id_pkey PRIMARY KEY (id)
);

CREATE TABLE etymology (
    id TEXT NOT NULL
    ,description TEXT
    ,entry_id BIGINT NOT NULL

    ,CONSTRAINT etymology_id_pkey PRIMARY KEY (id)
);

CREATE TABLE usage (
    id BIGINT NOT NULL GENERATED BY DEFAULT AS IDENTITY
    ,etymology_id TEXT NOT NULL

    ,CONSTRAINT usage_id_pkey PRIMARY KEY (id)
);

CREATE TABLE "group" (
    id BIGINT NOT NULL GENERATED BY DEFAULT AS IDENTITY
    ,description TEXT
    ,usage_id BIGINT NOT NULL

    ,CONSTRAINT group_id_pkey PRIMARY KEY (id)
);

CREATE TABLE definition (
    id BIGINT NOT NULL GENERATED BY DEFAULT AS IDENTITY
    ,text TEXT NOT NULL
    ,usage_id BIGINT
    ,group_id BIGINT

    ,CONSTRAINT definition_id_pkey PRIMARY KEY (id)
);

CREATE TABLE example (
    id BIGINT NOT NULL GENERATED BY DEFAULT AS IDENTITY
    ,text TEXT NOT NULL
    ,definition_id BIGINT NOT NULL

    ,CONSTRAINT example_id_pkey PRIMARY KEY (id)
);

-- 20230221194111_02_add_dictionary_language_fkeys.tx.sql
ALTER TABLE dictionary ADD CONSTRAINT dictionary_source_language_id_fkey FOREIGN KEY (source_language_id) REFERENCES language (id);

ALTER TABLE dictionary ADD CONSTRAINT dictionary_target_language_id_fkey FOREIGN KEY (target_language_id) REFERENCES language (id);

-- 20230221194111_03_add_entry_dictionary_fkeys.tx.sql
ALTER TABLE entry ADD CONSTRAINT entry_dictionary_id_fkey FOREIGN KEY (dictionary_id) REFERENCES dictionary (id);

-- 20230221194111_04_add_etymology_entry_fkeys.tx.sql
ALTER TABLE etymology ADD CONSTRAINT etymology_entry_id_fkey FOREIGN KEY (entry_id) REFERENCES entry (id);

-- 20230221194111_05_add_usage_etymology_fkeys.tx.sql
ALTER TABLE usage ADD CONSTRAINT usage_etymology_id_fkey FOREIGN KEY (etymology_id) REFERENCES etymology (id);

-- 20230221194111_06_add_group_usage_fkeys.tx.sql
ALTER TABLE "group" ADD CONSTRAINT group_usage_id_fkey FOREIGN KEY (usage_id) REFERENCES usage (id);

-- 20230221194111_07_add_definition_usage_fkeys.tx.sql
ALTER TABLE definition ADD CONSTRAINT definition_usage_id_fkey FOREIGN KEY (usage_id) REFERENCES usage (id);

-- 20230221194111_08_add_definition_group_fkeys.tx.sql
ALTER TABLE definition ADD CONSTRAINT definition_group_id_fkey FOREIGN KEY (group_id) REFERENCES "group" (id);

-- 20230221194111_09_add_example_definition_fkeys.tx.sql
ALTER TABLE example ADD CONSTRAINT example_definition_id_fkey FOREIGN KEY (definition_id) REFERENCES definition (id);

--[ mysql ]--

-- 20230221194111_01_tables.sql
CREATE TABLE language (
    id BIGINT AUTO_INCREMENT NOT NULL
    ,code VARCHAR(255) NOT NULL
    ,flag VARCHAR(255) NOT NULL

    ,PRIMARY KEY (id)
    ,CONSTRAINT language_code_key UNIQUE (code)
    ,CONSTRAINT language_flag_key UNIQUE (flag)
);

CREATE TABLE dictionary (
    id BIGINT AUTO_INCREMENT NOT NULL
    ,name VARCHAR(255) NOT NULL
    ,source_language_id BIGINT NOT NULL
    ,target_language_id BIGINT NOT NULL

    ,PRIMARY KEY (id)
);

CREATE TABLE entry (
    id BIGINT AUTO_INCREMENT NOT NULL
    ,created_at DATETIME DEFAULT CURRENT_TIMESTAMP
    ,updated_at DATETIME
    ,term VARCHAR(255) NOT NULL
    ,dictionary_id BIGINT

    ,PRIMARY KEY (id)
);

CREATE TABLE etymology (
    id VARCHAR(255) NOT NULL
    ,description VARCHAR(255)
    ,entry_id BIGINT NOT NULL

    ,PRIMARY KEY (id)
);

CREATE TABLE `usage` (
    id BIGINT AUTO_INCREMENT NOT NULL
    ,etymology_id VARCHAR(255) NOT NULL

    ,PRIMARY KEY (id)
);

CREATE TABLE `group` (
    id BIGINT AUTO_INCREMENT NOT NULL
    ,description VARCHAR(255)
    ,usage_id BIGINT NOT NULL

    ,PRIMARY KEY (id)
);

CREATE TABLE definition (
    id BIGINT AUTO_INCREMENT NOT NULL
    ,text VARCHAR(255) NOT NULL
    ,usage_id BIGINT
    ,group_id BIGINT

    ,PRIMARY KEY (id)
);

CREATE TABLE example (
    id BIGINT AUTO_INCREMENT NOT NULL
    ,text VARCHAR(255) NOT NULL
    ,definition_id BIGINT NOT NULL

    ,PRIMARY KEY (id)
);

-- 20230221194111_02_add_dictionary_source_language_id_fkey.sql
ALTER TABLE dictionary ADD CONSTRAINT dictionary_source_language_id_fkey FOREIGN KEY (source_language_id) REFERENCES language (id);

-- 20230221194111_03_add_dictionary_target_language_id_fkey.sql
ALTER TABLE dictionary ADD CONSTRAINT dictionary_target_language_id_fkey FOREIGN KEY (target_language_id) REFERENCES language (id);

-- 20230221194111_04_add_entry_dictionary_id_fkey.sql
ALTER TABLE entry ADD CONSTRAINT entry_dictionary_id_fkey FOREIGN KEY (dictionary_id) REFERENCES dictionary (id);

-- 20230221194111_05_add_etymology_entry_id_fkey.sql
ALTER TABLE etymology ADD CONSTRAINT etymology_entry_id_fkey FOREIGN KEY (entry_id) REFERENCES entry (id);

-- 20230221194111_06_add_usage_etymology_id_fkey.sql
ALTER TABLE `usage` ADD CONSTRAINT usage_etymology_id_fkey FOREIGN KEY (etymology_id) REFERENCES etymology (id);

-- 20230221194111_07_add_group_usage_id_fkey.sql
ALTER TABLE `group` ADD CONSTRAINT group_usage_id_fkey FOREIGN KEY (usage_id) REFERENCES `usage` (id);

-- 20230221194111_08_add_definition_usage_id_fkey.sql
ALTER TABLE definition ADD CONSTRAINT definition_usage_id_fkey FOREIGN KEY (usage_id) REFERENCES `usage` (id);

-- 20230221194111_09_add_definition_group_id_fkey.sql
ALTER TABLE definition ADD CONSTRAINT definition_group_id_fkey FOREIGN KEY (group_id) REFERENCES `group` (id);

-- 20230221194111_10_add_example_definition_id_fkey.sql
ALTER TABLE example ADD CONSTRAINT example_definition_id_fkey FOREIGN KEY (definition_id) REFERENCES definition (id);

--[ sqlserver ]--

-- 20230221194111_01_tables.sql
CREATE TABLE language (
    id BIGINT NOT NULL IDENTITY
    ,code NVARCHAR(255) NOT NULL
    ,flag NVARCHAR(255) NOT NULL

    ,CONSTRAINT language_id_pkey PRIMARY KEY (id)
    ,CONSTRAINT language_code_key UNIQUE (code)
    ,CONSTRAINT language_flag_key UNIQUE (flag)
);

CREATE TABLE dictionary (
    id BIGINT NOT NULL IDENTITY
    ,name NVARCHAR(255) NOT NULL
    ,source_language_id BIGINT NOT NULL
    ,target_language_id BIGINT NOT NULL

    ,CONSTRAINT dictionary_id_pkey PRIMARY KEY (id)
);

CREATE TABLE entry (
    id BIGINT NOT NULL IDENTITY
    ,created_at DATETIMEOFFSET DEFAULT CURRENT_TIMESTAMP
    ,updated_at DATETIMEOFFSET
    ,term NVARCHAR(255) NOT NULL
    ,dictionary_id BIGINT

    ,CONSTRAINT entry_id_pkey PRIMARY KEY (id)
);

CREATE TABLE etymology (
    id NVARCHAR(255) NOT NULL
    ,description NVARCHAR(255)
    ,entry_id BIGINT NOT NULL

    ,CONSTRAINT etymology_id_pkey PRIMARY KEY (id)
);

CREATE TABLE usage (
    id BIGINT NOT NULL IDENTITY
    ,etymology_id NVARCHAR(255) NOT NULL

    ,CONSTRAINT usage_id_pkey PRIMARY KEY (id)
);

CREATE TABLE [group] (
    id BIGINT NOT NULL IDENTITY
    ,description NVARCHAR(255)
    ,usage_id BIGINT NOT NULL

    ,CONSTRAINT group_id_pkey PRIMARY KEY (id)
);

CREATE TABLE definition (
    id BIGINT NOT NULL IDENTITY
    ,text NVARCHAR(255) NOT NULL
    ,usage_id BIGINT
    ,group_id BIGINT

    ,CONSTRAINT definition_id_pkey PRIMARY KEY (id)
);

CREATE TABLE example (
    id BIGINT NOT NULL IDENTITY
    ,text NVARCHAR(255) NOT NULL
    ,definition_id BIGINT NOT NULL

    ,CONSTRAINT example_id_pkey PRIMARY KEY (id)
);

-- 20230221194111_02_add_dictionary_language_fkeys.tx.sql
ALTER TABLE dictionary ADD CONSTRAINT dictionary_source_language_id_fkey FOREIGN KEY (source_language_id) REFERENCES language (id);

ALTER TABLE dictionary ADD CONSTRAINT dictionary_target_language_id_fkey FOREIGN KEY (target_language_id) REFERENCES language (id);

-- 20230221194111_03_add_entry_dictionary_fkeys.tx.sql
ALTER TABLE entry ADD CONSTRAINT entry_dictionary_id_fkey FOREIGN KEY (dictionary_id) REFERENCES dictionary (id);

-- 20230221194111_04_add_etymology_entry_fkeys.tx.sql
ALTER TABLE etymology ADD CONSTRAINT etymology_entry_id_fkey FOREIGN KEY (entry_id) REFERENCES entry (id);

-- 20230221194111_05_add_usage_etymology_fkeys.tx.sql
ALTER TABLE usage ADD CONSTRAINT usage_etymology_id_fkey FOREIGN KEY (etymology_id) REFERENCES etymology (id);

-- 20230221194111_06_add_group_usage_fkeys.tx.sql
ALTER TABLE [group] ADD CONSTRAINT group_usage_id_fkey FOREIGN KEY (usage_id) REFERENCES usage (id);

-- 20230221194111_07_add_definition_usage_fkeys.tx.sql
ALTER TABLE definition ADD CONSTRAINT definition_usage_id_fkey FOREIGN KEY (usage_id) REFERENCES usage (id);

-- 20230221194111_08_add_definition_group_fkeys.tx.sql
ALTER TABLE definition ADD CONSTRAINT definition_group_id_fkey FOREIGN KEY (group_id) REFERENCES [group] (id);

-- 20230221194111_09_add_example_definition_fkeys.tx.sql
ALTER TABLE example ADD CONSTRAINT example_definition_id_fkey FOREIGN KEY (definition_id) REFERENCES definition (id);
```

**Comment by maxdemaio** [comment]
Thank you so much for the help on this!!! I saw your YouTube presentation on `sq` and thought it'd be a great for this use case. Awesome work on `sq` and it's clearly very powerful 💪🏻

@Nickersoft I'll continue to move forward on this feature thanks to @bokwoon95's help!

**Comment by bokwoon95** [comment]
Happy to help. If you just want the SQL statements without the comments, use generateCmd.Results() instead. Example:

```go
var buf bytes.Buffer
generateCmd := &ddl.GenerateCmd{
	Dialect:   dialect,
	DirFS:     fsys,
	Filenames: []string{"main.go"},
}
files, _, err := generateCmd.Results()
if err != nil {
	fmt.Println(err)
	continue
}
for _, file := range files {
	if buf.Len() > 0 {
		buf.WriteString("\n")
	}
	buf.ReadFrom(file)
	file.Close()
}
fmt.Println(buf.String())
```
```sql
CREATE TABLE language (
    id BIGINT NOT NULL GENERATED BY DEFAULT AS IDENTITY
    ,code TEXT NOT NULL
    ,flag TEXT NOT NULL

    ,CONSTRAINT language_id_pkey PRIMARY KEY (id)
    ,CONSTRAINT language_code_key UNIQUE (code)
    ,CONSTRAINT language_flag_key UNIQUE (flag)
);

CREATE TABLE dictionary (
    id BIGINT NOT NULL GENERATED BY DEFAULT AS IDENTITY
    ,name TEXT NOT NULL
    ,source_language_id BIGINT NOT NULL
    ,target_language_id BIGINT NOT NULL

    ,CONSTRAINT dictionary_id_pkey PRIMARY KEY (id)
);

CREATE TABLE entry (
    id BIGINT NOT NULL GENERATED BY DEFAULT AS IDENTITY
    ,created_at TIMESTAMPTZ DEFAULT CURRENT_TIMESTAMP
    ,updated_at TIMESTAMPTZ
    ,term TEXT NOT NULL
    ,dictionary_id BIGINT

    ,CONSTRAINT entry_id_pkey PRIMARY KEY (id)
);

CREATE TABLE etymology (
    id TEXT NOT NULL
    ,description TEXT
    ,entry_id BIGINT NOT NULL

    ,CONSTRAINT etymology_id_pkey PRIMARY KEY (id)
);

CREATE TABLE usage (
    id BIGINT NOT NULL GENERATED BY DEFAULT AS IDENTITY
    ,etymology_id TEXT NOT NULL

    ,CONSTRAINT usage_id_pkey PRIMARY KEY (id)
);

CREATE TABLE "group" (
    id BIGINT NOT NULL GENERATED BY DEFAULT AS IDENTITY
    ,description TEXT
    ,usage_id BIGINT NOT NULL

    ,CONSTRAINT group_id_pkey PRIMARY KEY (id)
);

CREATE TABLE definition (
    id BIGINT NOT NULL GENERATED BY DEFAULT AS IDENTITY
    ,text TEXT NOT NULL
    ,usage_id BIGINT
    ,group_id BIGINT

    ,CONSTRAINT definition_id_pkey PRIMARY KEY (id)
);

CREATE TABLE example (
    id BIGINT NOT NULL GENERATED BY DEFAULT AS IDENTITY
    ,text TEXT NOT NULL
    ,definition_id BIGINT NOT NULL

    ,CONSTRAINT example_id_pkey PRIMARY KEY (id)
);

ALTER TABLE dictionary ADD CONSTRAINT dictionary_source_language_id_fkey FOREIGN KEY (source_language_id) REFERENCES language (id);

ALTER TABLE dictionary ADD CONSTRAINT dictionary_target_language_id_fkey FOREIGN KEY (target_language_id) REFERENCES language (id);

ALTER TABLE entry ADD CONSTRAINT entry_dictionary_id_fkey FOREIGN KEY (dictionary_id) REFERENCES dictionary (id);

ALTER TABLE etymology ADD CONSTRAINT etymology_entry_id_fkey FOREIGN KEY (entry_id) REFERENCES entry (id);

ALTER TABLE usage ADD CONSTRAINT usage_etymology_id_fkey FOREIGN KEY (etymology_id) REFERENCES etymology (id);

ALTER TABLE "group" ADD CONSTRAINT group_usage_id_fkey FOREIGN KEY (usage_id) REFERENCES usage (id);

ALTER TABLE definition ADD CONSTRAINT definition_usage_id_fkey FOREIGN KEY (usage_id) REFERENCES usage (id);

ALTER TABLE definition ADD CONSTRAINT definition_group_id_fkey FOREIGN KEY (group_id) REFERENCES "group" (id);

ALTER TABLE example ADD CONSTRAINT example_definition_id_fkey FOREIGN KEY (definition_id) REFERENCES definition (id);
```

And I just saw you have to generate INSERT queries as well. You can use the same table structs to build your queries.

```go
type LANGUAGE struct {
	sq.TableStruct
	ID   sq.NumberField `ddl:"type=BIGINT primarykey auto_increment autoincrement identity"`
	CODE sq.StringField `ddl:"notnull unique"`
	FLAG sq.StringField `ddl:"notnull unique"`
}

languages := [][2]string{
	{"abc", "def"},
	{"ghi", "jkl"},
	{"mno", "pqr"},
}
for _, language := range languages {
	l := sq.New[LANGUAGE]("")
	insertQuery := sq.
		InsertInto(l).
		Columns(l.CODE, l.FLAG).
		Values(sq.Literal(language[0]), sq.Literal(language[1]))
	query, _, err := sq.ToSQL(dialect, insertQuery, nil)
	if err != nil {
		fmt.Println(err)
		continue
	}
	fmt.Println(query + ";")
}
```
```sql
INSERT INTO language (code, flag) VALUES ('abc', 'def');
INSERT INTO language (code, flag) VALUES ('ghi', 'jkl');
INSERT INTO language (code, flag) VALUES ('mno', 'pqr');
```

**Comment by maxdemaio** [comment]
@bokwoon95 this was exactly what we needed! After following by your examples, things seem to be working 😁 

With the command `go run odict.go d -f postgres examples/example1.odict examples/output.sql` it will generate an `output.sql` file that will seed a database. Also, this will work with every dialect by passing in `sqlite`, `mysql`, and `sqlserver` accordingly.

<img width="980" alt="image" src="https://user-images.githubusercontent.com/44373521/220813866-59c5b0c3-63ea-4282-876f-20e3dc65dc5f.png">

I will hopefully be getting this work into PR soon 👍🏻
